resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+        `{V_1, V_2, ...}` be the _grounded_ constraint solution for the set of
+        type variables `{X_1, X_2, ...}` with respect to the constaint set `C`,
+        with partial solution `{U_1, U_2, ...}`. Replace `{U_1, U_2, ...}` with
+        `{V_1, V_2, ...}`. _This step is known as "upwards inference", since it
+        has the effect of passing type information __up__ the syntax tree from
+        the static type of arguments to the inferred type arguments of the
+        invocation._
+
+- _Note that at this point, all entries in `{U_1, U_2, ...}` have been obtained
+  either from a _grounded_ constraint solution or from explicit types in the
+  source code, so they are guaranteed to be proper types, not type schemas._
+
+- For each `e_i`:
+
+  - Let `K_i` be the result of substituting `{U_1, U_2, ...}` for `{X_1, X_2,
+    ...}` in `P_i`. _Note that this is now guaranteed to be a proper type, not a
+    type schema._
+
+  - Let `m_i` be the result of performing coercion of `m_i_preliminary` to type
+    `K_i`. _This ensures that the invocation is sound._
+
+- Finally, let `R` be the result of substituting `{U_1, U_2, ...}` for `{X_1,
+  X_2, ...}` in `R_F`.
+
+### Argument Partitioning
+
+In the algorithm above, the argument partitioning works as follows. First there
+is a _dependency analysis_ phase, in which type inference decides which
+invocation arguments might benefit from being type inferred before other
+arguments, and then a _stage selection_ phase, in which type inference
+partitions the arguments into stages.
+
+#### Dependency analysis
+
+First, a dependency graph is formed among the invocation arguments `e_i` based
+on the following rule: there is a dependency edge from argument `e_i` to
+argument `e_j` if and only if the type of the invocation target is generic, and
+the following relationship exists among `e_i`, `e_j`, and at least one of the
+formal type parameters `X_k`:
+
+1. `e_i` is a _&lt;functionExpression&gt;_ enclosed in zero or more parentheses,
+
+2. AND `P_i` is a function type,
+
+3. AND `X_k` is a free variable in the type of at least one of the parameters of
+   `P_i`,
+
+4. AND the corresponding parameter in `e_i` does not have a type annotation,
+
+5. AND EITHER:
+
+   * `P_j` is a function type, and `X_k` is a free variable in the return type
+     of `P_j`
+
+   * OR `P_j` is _not_ a function type, and `X_k` is a free variable in `P_j`.
+
+_The idea here is that we're trying to draw a dependency edge from `e_i` to
+`e_j` in precisely those cirumstances in which there is likely to be a benefit
+to performing a round of horizontal inference after type inferring `e_j` and
+before type inferring `e_i`._
+
+_Note that if `F` is `∅`, then the resulting graph has no edges._
+
+_So, for example, if the invocation in question is this:_
+
+```dart
+f((t, u) { ... } /* A */, () { ... } /* B */, (v) { ... } /* C */, (u) { ... } /* D */)
```

Comment:
This example is very hard to parse as written.  Consider formatting it with arguments on different lines parallel to the declaration below?

Reply:
Haha, this line was copied wholesale from https://github.com/dart-lang/language/blob/main/accepted/2.18/horizontal-inference/feature-specification.md, and no one complained about it there 🙂

But you have a good point. Reformatted.